src/images/images.service.ts

```diff
@@ -68,6 +70,75 @@ export class ImagesService {
       );
     });
 
-    return { data };
+    return { images: data };
+  }
+
+  async searchImages(
+    searchQuery = '',
+    page = 1,
+    perPage: number,
+    userId?: number | string,
+  ): Promise<{
+    count: number;
+    page: number;
+    perPage: number;
+    images: Image[];
+  }> {
+    const params = searchQuery.split(' ');
+    const arr = permutationsOfArray(params.slice(0, 3));
+    const arrOfPromises: Promise<Image[]>[] = [];
+    const { limit, offset } = paginate(page, perPage);
+    const watermark = await this.watermarksRepository.findOneBy({
+      isDefault: true,
+    });
+
+    arr.forEach((query) => {
+      arrOfPromises.push(
+        this.imagesRepository.find({
+          select: {
+            id: true,
+            name: true,
+            tags: true,
+            createdAt: true,
+            updatedAt: true,
+            user: { id: true, displayName: true },
+          },
+          where: {
+            tags: Like(query),
+            user: Like(userId ? userId : '%'),
+            isApproved: true,
+          },
+          relations: ['user'],
+        }),
+      );
+    });
+
+    let result = [];
+    await Promise.all(arrOfPromises).then((permutedOutput) => {
+      permutedOutput.forEach((arrOfImages) => {
+        result = result.filter((image) => {
+          return !arrOfImages.some((photo) => {
+            return image.id === photo.id;
+          });
+        });
+        result = result.concat(arrOfImages);
+      });
+    });
+
+    result.forEach((image) => {
+      image.path = makeUrlPath([
+        'images',
+        watermark.id,
+        process.env.BASE_THUMBNAIL_SIZE,
+        image.name,
+      ]);
+    });
+
+    const count = result.length;
+    result = result
```

Comment:
Related to the previous comment - yeah, with this approach there is no real db pagination, we are taking out bunch of records and then cut them with this slice. We can improve this, I'll see to send you some examples